### DISCUSSION TITLE
Poetry as packager and dependency manager

### DISCUSSION THREAD

**Discussion by Alec4r** [eduNEXT/tvm#4] [Ideas]
# What is [Poetry](https://python-poetry.org/)
Poetry is a tool for dependency management and packaging in Python. It allows you to declare the libraries your project depends on and it will manage (install/update) them for you.

# Why we must use poetry?
## Dependency management and Packaging
- **poetry.lock**: This file will be updated each that updates a dependency, and when a new developer wants to develop a new feature will just need to run `poetry install` to get the dependencies with which was developed TVM.
- **pyproject.toml**: This file replaces the old setup.py file, here you can find package info like name, version, description, authors, ...etc. Also, you can find dependencies by environment or scripts (python commands).

## Python environments
Also with poetry, you won't need to create a virtualenv, since with poetry you can run commands using `poetry run <your command>` or enable a virtual python environment using `poetry shell`, if you want use a specific python environment you can do it with `poetry env use <python path or python-version>`

## Build & Publish
- [build](https://python-poetry.org/docs/cli#build): `poetry build`
- [publish](https://python-poetry.org/docs/cli#publish): `poetry publish`

**Comment by black-redoc** [comment]
I was looking for information about poetry and I found several approaches:

- Manage virtualenvs
- Manage python interpreter (python versions)
- Manage several development tools (flake8, pytest, etc.)
- Make the build easily and it's agnostic of setuptools
- Upload the build to pypi easily

This package has several uses which might help us to do the development quick and easy.

**Comment by mariajgrimaldi** [comment]
I'd like to know why poetry instead of the old but well-known pip? :thinking:. There're some intriguing comments on [this blog post](https://news.ycombinator.com/item?id=26093926) about Poetry for anyone interested.

**Reply by Alec4r** [reply]
It is just another option, It's newer and better than pip-tools __for me__ , because with pip tools you need to call `pip freeze requirements.txt` all the time, if you forget it you will have problems, also, if you want to set "dev requirements" and install "dev requirements" you have 2 options "dev.txt" or "extras_require[dev] in setup.py", all this is too boring for me. 

## In poetry
```bash
poetry add requests
poetry add --dev pytest
````
This will update automatically file versions and file .lock

```bash
poetry install
poetry install --no-dev
```

**Reply by Alec4r** [reply]
Python is the programming language that doesn't use a good dependency manager/packager by default, poetry is like npm to javascript or rubygems to ruby.
